<p>
    <a name="R1" id="R1"></a>1. Hospital, A., Faustino, I.,  Collepardo-Guevara, R., Gonzalez, C., Gelpi, J.L. and Orozco, M. (2013) NAFlex:  a web server for the study of nucleic acid flexibility. <em>Nucleic Acids Research</em>, <strong>41</strong>,  W47-W55.<br />
</p>
<p>
    <a name="R2" id="R2"></a>2. Michaud-Agrawal,  N., Denning, E.J., Woolf, T.B. and Beckstein, O. (2011) Software News and  Updates MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics  Simulations. <em>Journal of Computational  Chemistry</em>, <strong>32</strong>, 2319-2327.<br />
</p>
<p>
    <a name="R3" id="R3"></a>3. Lakshman, A. and  Malik, P. (2010), SIGOPS Oper. Syst. Rev, Vol. 44, pp. 35-40.<br />
</p>
<p>
    <a name="R4" id="R4"></a>4. Hernandez, R.,  Cugnasco, C., Becerra, Y., Torres, J. and Ayguade, E. (2015), Proceedings of  the 23rd Euromicro International Conference on Parallel, Distributed, and  Network-Based Processing, pp. 288-295.<br />
</p>
<p>
    <a name="R5" id="R5"></a>5. Hernandez, R.,  Becerra, Y., Torres, J. and Ayguade, E. (2015), Proceedings of the  International Conference on Computational Science, ICCS 2015, pp. 2822-2826.
</p>
<p><a name="R6" id="R6"></a>6.	Ison, J., Kalas, M., Jonassen, I., Bolser, D., Uludag, M., McWilliam, H., Malone, J., Lopez, R., Pettifer, S. and Rice, P. (2013) EDAM: an ontology of bioinformatics operations, types of data and identifiers, topics and formats. Bioinformatics, 29, 1325-1332.  </p>
<p><a name="R7" id="R7"></a>7.	Ashburner, M., Ball, C.A., Blake, J.A., Botstein, D., Butler, H., Cherry, J.M., Davis, A.P., Dolinski, K., Dwight, S.S., Eppig, J.T. et al. (2000) Gene Ontology: tool for the unification of biology. Nature Genetics, 25, 25-29. </p>
<p><a name="R8" id="R8"></a>8.	Smith, B., Ashburner, M., Rosse, C., Bard, J., Bug, W., Ceusters, W., Goldberg, L.J., Eilbeck, K., Ireland, A., Mungall, C.J. et al. (2007) The OBO Foundry: coordinated evolution of ontologies to support biomedical data integration. Nat Biotechnol, 25, 1251-1255.  </p>
<p><a name="R9" id="R9"></a>9.	Hastings, J., de Matos, P., Dekker, A., Ennis, M., Harsha, B., Kale, N., Muthukrishnan, V., Owen, G., Turner, S., Williams, M. et al. (2013) The ChEBI reference database and ontology for biologically relevant chemistry: enhancements for 2013. Nucleic Acids Research, 41, D456-D463.  </p>
<p><a name="R10" id="R10"></a>10.	Pasi, M., Maddocks, J.H., Beveridge, D., Bishop, T.C., Case, D.A., Cheatham, T., III, Dans, P.D., Jayaram, B., Lankas, F., Laughton, C. et al. (2014) mu ABC: a systematic microsecond molecular dynamics study of tetranucleotide sequence effects in B-DNA. Nucleic Acids Research, 42, 12272-12283.  </p>
<p><a name="R11" id="R11"></a>11.	Dans, P.D., Faustino, I., Battistini, F., Zakrzewska, K., Lavery, R. and Orozco, M. (2014) Unraveling the sequence-dependent polymorphic behavior of d(CpG) steps in B-DNA. Nucleic Acids Research, 42, 11304-11320.  </p>
<p><a name="R12" id="R12"></a>12.	Berman, H.M., Battistuz, T., Bhat, T.N., Bluhm, W.F., Bourne, P.E., Burkhardt, K., Iype, L., Jain, S., Fagan, P., Marvin, J. et al. (2002) The Protein Data Bank. Acta Crystallographica Section D-Biological Crystallography, 58, 899-907. </p>
<p><a name="R13" id="R13"></a>13.	Blanchet, C., Pasi, M., Zakrzewska, K. and Lavery, R. (2011) CURVES plus web server for analyzing and visualizing the helical, backbone and groove parameters of nucleic acid structures. Nucleic Acids Research, 39, W68-W73. </p>
<p><a name="R14" id="R14"></a>14.	Roe, D.R. and Cheatham, T.E., III. (2013) PTRAJ and CPPTRAJ: Software for Processing and Analysis of Molecular Dynamics Trajectory Data. Journal of Chemical Theory and Computation, 9, 3084-3095.  </p>
<p><a name="R15" id="R15"></a>15.	Pronk, S., Pall, S., Schulz, R., Larsson, P., Bjelkmar, P., Apostolov, R., Shirts, M.R., Smith, J.C., Kasson, P.M., van der Spoel, D. et al. (2013) GROMACS 4.5: a high-throughput and highly parallel open source molecular simulation toolkit. Bioinformatics, 29, 845-854.  </p>
<p><a name="R16" id="R16"></a>16.	Meyer, T., Ferrer-Costa, C., Perez, A., Rueda, M., Bidon-Chanal, A., Luque, F.J., Laughton, C.A. and Orozco, M. (2006) Essential dynamics: A tool for efficient trajectory compression and management. Journal of Chemical Theory and Computation, 2, 251-258. </p>
<p><a name="R17" id="R17"></a>17.	Camps, J., Carrillo, O., Emperador, A., Orellana, L., Hospital, A., Rueda, M., Cicin-Sain, D., D'Abramo, M.,  Gelpi, J.L. and Orozco, M. (2009) FlexServ: an integrated tool for the analysis of protein flexibility. Bioinformatics, 25, 1709-1710.  </p>
<p><a name="R18" id="R18"></a>18.	Humphrey, W., Dalke, A. and Schulten, K. (1996) VMD: Visual molecular dynamics. Journal of Molecular Graphics & Modelling, 14, 33-38. </p>
